src/KurrentDB.Api.V2/Modules/Streams/StreamsService.cs

```diff
@@ -0,0 +1,265 @@
+            var callback = new AppendRecordsCallback(requests: Requests);
+
+            // not yet because of all the validations
+            // in the constructor that are helpful.
+            // should actually transform it into tests.
+            // var command = new WriteEvents(
+            //     callback,
+            //     Streams.ToImmutable(),
+            //     Revisions.ToImmutable(),
+            //     Events.ToImmutable(),
+            //     Indexes.ToImmutable(),
+            //     user,
+            //     cancellationToken
+            // );
+
+            var cid = Guid.NewGuid();
+            var command = new WriteEvents(
+                internalCorrId: cid,
+                correlationId: cid,
+                envelope: callback,
+                requireLeader: true,
+                eventStreamIds: Streams.ToImmutable(),
+                expectedVersions: Revisions.ToImmutable(),
+                events: Events.ToImmutable(),
+                eventStreamIndexes: Indexes.ToImmutable(),
+                user: user,
+                tokens: null,
+                cancellationToken: cancellationToken
+            );
+
+            Publisher.Publish(message: command);
+
+            return callback.WaitForReply;
+        }
+    }
+
+    class AppendRecordsCallback(IndexedSet<AppendRequest> requests)
+        : ApiCallback<IndexedSet<AppendRequest>, AppendSessionResponse, RpcException>(requests, "streams.append-session") {
+        protected override bool SuccessPredicate(Message message, IndexedSet<AppendRequest> requests) =>
+            message is WriteEventsCompleted { Result: OperationResult.Success };
+
+        protected override AppendSessionResponse MapToApiResponse(Message message, IndexedSet<AppendRequest> requests) {
+            var completed = (WriteEventsCompleted)message;
+            var output    = new List<AppendResponse>();
+
+            for (var i = 0; i < completed.LastEventNumbers.Length; i++)
+                output.Add(new() {
+                    Stream         = requests.ElementAt(i).Stream,
+                    StreamRevision = completed.LastEventNumbers.Span[i]
+                });
+
+            return new() {
+                Output   = { output },
+                Position = completed.CommitPosition
+            };
+        }
+
+        protected override RpcException MapToApiError(Message message, IndexedSet<AppendRequest> requests) {
+            return message switch {
+                WriteEventsCompleted completed => completed.Result switch {
+                    // because we can append to soft deleted streams, StreamDeleted means "hard deleted" AKA tombstoned
+                    OperationResult.StreamDeleted  => ApiErrors.StreamTombstoned(requests.ElementAt(completed.FailureStreamIndexes.Span[0]).Stream),
+                    OperationResult.PrepareTimeout => ApiErrors.OperationTimeout(completed.Message),
```

Comment:
does this line need to be removed? as it's said below that `OperationResult.PrepareTimeout` is not possible in this context